nbest rescoring with k2

glynpu opened the pull request:
Performance on librispeech **_test_clean_** :
With batch_size=5, decoding duration is 665 seconds.

### WER
|method|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|---|
|espnet baseline 70d16d210cdce28e71b7892a0ec96eaaa6474d64 |2620 | 52576 | 98.1 | 1.7 | 0.2 | 0.3 | 2.1 | 26.2
|nbest_rescoring with k2|2620|52576|98.0|1.8|0.2|0.3|2.3|27.5|

sw005320 commented on espnet2/bin/asr_inference_k2.py:
```diff
@@ -519,6 +670,16 @@ def get_parser():
         help="The model path of sentencepiece. "
         "If not given, refers from the training args",
     )
+    group.add_argument("--is_ctc_decoding", type=str2bool, default=True)
```
`ctc_decoding` does a bit ambiguous to me.
ctc_viterbi_decoding? ctc_frame_sync_decoding?

glynpu replied:
"ctc_decoding" means the decoding graph only contains "a CTC topology"  and is NOT " TLG" decoding.
Shall I change "--is_ctc_decoding" to "--use_ctc_graph"?

sw005320 replied:
I see.
Then, I think ctc decoding is fine. 
Could you add some clarifications in the help?

sw005320 commented on espnet2/fst/lm_rescore.py:
```diff
@@ -0,0 +1,187 @@
+from typing import List
```
ditto.
please increase the test coverage.

glynpu replied:
fixed. Now it's covered by ci/test_integration_espnet2.sh.

sw005320 commented on egs2/TEMPLATE/asr1/asr.sh:
```diff
@@ -94,6 +94,34 @@ num_splits_asr=1           # Number of splitting for lm corpus.
 
 # Decoding related
 use_k2=false      # Whether to use k2 based decoder
+export is_ctc_decoding=true
```
Is it possible to add some of them in the inference config?

glynpu replied:
Good point. 
I will move most of these options to inference config file.
And several of them may remain into asr.sh to construct inference_tag.

Now **_export_** is used because shellcheck consided them as "unused varialbe" and ci fails. Here is an example log:
```
In egs2/librispeech/asr1/asr.sh line 106:
nnlm_weight="1.0"
^---------^ SC2034: nnlm_weight appears unused. Verify use (or export if used externally).
```
Though they are definitely used in following **_for loop_** to [construct inference_tag](https://github.com/espnet/espnet/blob/2e8c6cdad4cff64ae849a8af21d99099360490f5/egs2/TEMPLATE/asr1/asr.sh#L440).
```
    if "${use_k2}"; then
      inference_tag+="_use_k2"
      for flag in ${inference_tag_flags_k2}; do
        inference_tag+="_${flag}_${!flag}"
      done
    fi
```

glynpu replied:
fixed by add conf/decode_asr_transformer_with_k2.yaml.

Results are also added into README.md

sw005320 commented on espnet2/fst/lm_rescore.py:
```diff
@@ -0,0 +1,187 @@
+from typing import List
+from typing import Tuple
```
@takaaki-hori, could you check this part?

sw005320 commented on espnet2/lm/espnet_model.py:
```diff
@@ -1,4 +1,5 @@
 from typing import Dict
```
ditto
Please increase the test coverage.

glynpu replied:
fixed. Now it's coveraged by ci/test_integration.sh.

sw005320 reviewed (commented):
- It would be great if you prepare a config file in https://github.com/espnet/espnet/tree/master/egs2/librispeech/asr1/conf and add some results to https://github.com/espnet/espnet/blob/master/egs2/librispeech/asr1/README.md with the corresponding config file

glynpu commented on espnet2/asr/espnet_model.py:
```diff
@@ -267,6 +267,68 @@ def _extract_feats(
             feats, feats_lengths = speech, speech_lengths
         return feats, feats_lengths
 
+    def nll(
```
fixed. Now it's covered by ci/test_integration.sh.

ShigekiKarita replied:
can you add docstrings in nll and batchfy_nll about expected shapes (see `forward`) and nll means negative log likelihood.

glynpu replied:
fixed by adding doc docstrings.

glynpu commented:
@sw005320 Could you help to re-start the failed ci workflow.  
It failed because of "connection time-out" when installing espnet.

sw005320 commented:
@glynpu, could you fix the conflicts?
@ShigekiKarita, could you also check this PR?

It’s getting good shape. I’ll merge it soon.

ShigekiKarita commented:
Great PR! Can you merge our master branch? It recently changed coverage calc. #3600

ShigekiKarita commented on ci/test_integration.sh:
```diff
@@ -209,6 +209,17 @@ echo "==== feats_type=raw, token_types=bpe, model_conf.extract_feats_in_collect_
     --feats_normalize "utterance_mvn" --lm-args "--max_epoch=1" --python "${python}" \
     --asr-args "--model_conf extract_feats_in_collect_stats=false --max_epoch=1"
+
+echo "==== use_k2, num_paths > nll_batch_size, feats_type=raw, token_types=bpe, model_conf.extract_feats_in_collect_stats=False, normalize=utt_mvn ==="
```
FYI: now you should move these tests to ci/test_integration_espnet2.sh because of #3599.

glynpu replied:
fixed by moving it to ci/test_integration_espnet2.sh.

ShigekiKarita commented on espnet2/bin/asr_inference_k2.py:
```diff
@@ -1,5 +1,6 @@
 #!/usr/bin/env python3
 import argparse
+import datetime
```
see https://github.com/espnet/espnet/blob/master/test/espnet2/bin/test_k2_asr_inference.py

glynpu replied:
fixed. Besides test_k2_asr_inference.py, it's also coveraged by ci/test_integration_espnet2.sh.

glynpu commented:
> @glynpu, could you fix the conflicts?

Sure. Now conflicts are resolved by rebased on current master.
@sw005320  Could you help to restart the ci which are failed during installing espnet.

glynpu commented:
> Great PR! Can you merge our master branch? It recently changed coverage calc. #3600

@ShigekiKarita Sure. Now this pr is rebased on current master.

glynpu commented on espnet2/bin/asr_inference_k2.py:
```diff
@@ -220,27 +278,122 @@ def __call__(
         dense_fsa_vec = k2.DenseFsaVec(logp_encoder_output, supervision_segments)
```
fixed.

glynpu commented on espnet2/bin/asr_inference_k2.py:
```diff
@@ -148,10 +181,23 @@ def __init__(
+                build_ctc_topo(list(range(len(token_list))))
+            )
+
+        self.decode_graph = self.decode_graph.to(device)
```
fixed.

glynpu commented on espnet2/fst/lm_rescore.py:
```diff
@@ -0,0 +1,190 @@
+    del inverted_lats.aux_labels
+    inverted_lats = k2.arc_sort(inverted_lats)
+
+    am_path_lats = _intersect_device(
```
fixed.

glynpu commented:
> can you work on it given the feedback?

Now all the reviews are fixed, I think. 
There seems exists some ci errors, I will fix it tomorrow.

> We should have some sync.

Here are my recent results of chunk-streming model. 
I am facing some convering issues. Still working on it.
https://github.com/k2-fsa/icefall/pull/89

sw005320 commented:
@brianyan918 and @pengchengguo, could you check the update made by @glynpu?

pengchengguo reviewed (approved):
Sorry, I am not familiar with the lm score part. Other parts are good enough for me.

sw005320 commented:
It seems that we still have an issue in the k2 integration test.

glynpu commented:
> It seems that we still have an issue in the k2 integration test.

Maybe just restart the ci will solve problems. Because some ci failed during espnet installation because network connection and some failed at the place I didn't modify at all.

sw005320 commented:
OK, I'm rerunning it.
The error comes from the k2 integration test (other tests seem to be no problem) https://github.com/espnet/espnet/runs/4166114139?check_suite_focus=true#step:18:785
So, you can also reproduce it in your local machine (https://github.com/espnet/espnet/blob/master/CONTRIBUTING.md#5-integration-testing)

Also, we are currently touching on the installation process for some improvements, and it would be better to frequently merge the master.

glynpu commented:
With set fast-fail=false, a failed ci will not cancel other unfinished cis.

I made [a pr to my own master](https://github.com/glynpu/espnet/pull/35) with fast-fail=false.
Some ci **successfully** finished as expected while others still failed.

I am rerunning the failed ones to see what will happen.

kamo-naoyuki commented:
Sorry, I changed the k2 version used in CI from v1.3 to v1.6, which passed the unit test.
https://github.com/espnet/espnet/blob/ff870699fafabd13f3a10610a78cb2da3e386a64/tools/installers/install_k2.sh#L5-L6

Is this the reason for the CI error?

(BTW, v1.6 doesn't support pytorch>=1.9.1 and the latest k2, v1.10, support it, so it's better to make it compatible to use the latest pytorch)

glynpu commented:
> Sorry, I changed the k2 version used in CI from v1.3 to v1.6, which passed the unit test.
> 
> https://github.com/espnet/espnet/blob/ff870699fafabd13f3a10610a78cb2da3e386a64/tools/installers/install_k2.sh#L5-L6
> 
> Is this the reason for the CI error?
> 
> (BTW, v1.6 doesn't support pytorch>=1.9.1 and the latest k2, v1.10, support it, so it's better to make it compatible to use the latest pytorch)

I see.  Some apis are removed in from v1.6.
I changed the apis name to make this pr compatible to latest k2.
Now it's being tested and will be summited if wer and speed is consistent with previous results.

glynpu commented:
@sw005320 @kamo-naoyuki  Needs Help! 

torch1.3 will NOT be supported by current latest k2(v1.10) and future versions.
https://github.com/k2-fsa/k2/issues/822
So for ci check of this pr:
Solution 1: always use latest k2 and skip k2 related checks if torch==1.3.
Solution 2: use older versions of k2 to compatible with torch==1.3

Which should I do?

kamo-naoyuki commented:
> Solution 1: always use latest k2 and skip k2 related checks if torch==1.3.

I created a block for exception in install_k2.sh:
https://github.com/espnet/espnet/blob/c6139d5f0dc79fc4676fd908bbce05486a6672e1/tools/installers/install_k2.sh#L89-L121

Please modify it.

glynpu commented:
@kamo-naoyuki Could you give me some suggestions on how to fix this ci error? it failed at "test shell" step, I think didn't modify that.
https://github.com/glynpu/espnet/runs/4191386546?check_suite_focus=true

kamo-naoyuki commented:
> @kamo-naoyuki Could you give me some suggestions on how to fix this ci error? it failed at "test shell" step, I think didn't modify that. https://github.com/glynpu/espnet/runs/4191386546?check_suite_focus=true

Did you merge the current master? It has been fixed.

glynpu commented:
> > @kamo-naoyuki Could you give me some suggestions on how to fix this ci error? it failed at "test shell" step, I think didn't modify that. https://github.com/glynpu/espnet/runs/4191386546?check_suite_focus=true
> 
> Did you merge the current master? It has been fixed.

Thanks! Solved with rebased on current master.

glynpu commented:
@kamo-naoyuki Now all ci are fixed except test_debian9. Here is the log https://github.com/glynpu/espnet/runs/4197176742?check_suite_focus=true.

Could you give me some suggestions to fix it?

glynpu commented:
> Also, we are currently touching on the installation process for some improvements, and it would be better to frequently merge the master.

@sw005320 I also changed fairest installation in this pr.  As it will uninstall pytorch==1.8.0 and install a new pytorch==1.10.0, then ci fails. 
As this log:https://github.com/glynpu/espnet/runs/4197248694?check_suite_focus=true

Related modification in this pr:
```
    # FairSeq Commit id when making this PR: `commit 313ff0581561c7725ea9430321d6af2901573dfb`
-     git clone --depth 1 https://github.com/pytorch/fairseq.git
+     git clone https://github.com/pytorch/fairseq.git
+     if ! "$(pt_plus 1.10.0)" ; then
+        # Note(LiyongGuo): The latest fairseq will automatically uninstall
+         # the existing pytorch==1.8.0 and install pytorch==1.10.0,
+        # which cause an error of test_centos7 ci.
+         cd ./fairseq
+         git reset --hard 313ff0581561c7725ea9430321d6af2901573dfb
+         cd ../
+     fi
```

kamo-naoyuki commented:
> I also changed fairest installation in this pr. As it will uninstall pytorch==1.8.0 and install a new 

Please keep this as it is. This will be fixed by torchaudio side.

glynpu commented:
> > I also changed fairest installation in this pr. As it will uninstall pytorch==1.8.0 and install a new
> 
> Please keep this as it is. This will be fixed by torchaudio side.

So just ignore the test_centos7 ci error? 
This error also exist in current expense/master. https://github.com/espnet/espnet/runs/4186173742?check_suite_focus=true

kamo-naoyuki commented:
Yes, we are ignoring it now.

kamo-naoyuki commented:
> Here is the check with no modification, as you see, latest k2 is supported and all ci success except test_debian9. [glynpu#36 (comment)](https://github.com/glynpu/espnet/pull/36#partial-pull-merging)

We can't solve this issue in a straightforward way, but this is not serious and you don't need to fix it. I'll fix it somehow, so please ignore it in this PR for now.

glynpu commented:
>We can't solve this issue in a straightforward way, but this is not serious and you don't need to fix it. I'll fix it somehow, so please ignore it in this PR for now.

Is there anything I should do about this pr now? I think I have fixed some reviews.

Following is the performance with k2 v1.10, I think the result is as expected.
results with k2 v1.10.0
###WER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|                                                                                   
|---|---|---|---|---|---|---|---|---|                                                                                          
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_is_ctc_decoding_true_use_nbest_rescoring_false/dev_clean|2703|54402|97.5|2.2|0.3|0.3|2.7|33.8|                                                               
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_is_ctc_decoding_true_use_nbest_rescoring_false/dev_other|2864|50948|93.6|5.8|0.6|0.8|7.1|53.5|                                                               
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_is_ctc_decoding_true_use_nbest_rescoring_false/test_clean|2620|52576|97.3|2.4|0.2|0.3|3.0|34.9|                                                              
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_is_ctc_decoding_true_use_nbest_rescoring_false/test_other|2939|52343|93.8|5.6|0.6|0.8|7.0|54.1|                                                              
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_is_ctc_decoding_true_use_nbest_rescoring_true/dev_clean|2703|54402|98.2|1.6|0.2|0.2|2.0|26.6|                                                                
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_is_ctc_decoding_true_use_nbest_rescoring_true/dev_other|2864|50948|95.1|4.3|0.5|0.5|5.4|42.7|                                                                
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_is_ctc_decoding_true_use_nbest_rescoring_true/test_clean|2620|52576|98.0|1.8|0.2|0.3|2.3|27.5|                                                               
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_is_ctc_decoding_true_use_nbest_rescoring_true/test_other|2939|52343|95.2|4.2|0.6|0.5|5.3|44.7| 


###CER

|dataset|Snt|Wrd|Corr|Sub|Del|Ins|Err|S.Err|                                                                                   
|---|---|---|---|---|---|---|---|---|                                                                                          
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_is_ctc_decoding_true_use_nbest_rescoring_false/dev_clean|2703|54402|97.5|2.2|0.3|0.3|2.7|33.8|                                                               
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_is_ctc_decoding_true_use_nbest_rescoring_false/dev_other|2864|50948|93.6|5.8|0.6|0.8|7.1|53.5|                                                               
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_is_ctc_decoding_true_use_nbest_rescoring_false/test_clean|2620|52576|97.3|2.4|0.2|0.3|3.0|34.9|                                                              
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_is_ctc_decoding_true_use_nbest_rescoring_false/test_other|2939|52343|93.8|5.6|0.6|0.8|7.0|54.1|                                                              
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_is_ctc_decoding_true_use_nbest_rescoring_true/dev_clean|2703|54402|98.2|1.6|0.2|0.2|2.0|26.6|                                                                
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_is_ctc_decoding_true_use_nbest_rescoring_true/dev_other|2864|50948|95.1|4.3|0.5|0.5|5.4|42.7|                                                                
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_is_ctc_decoding_true_use_nbest_rescoring_true/test_clean|2620|52576|98.0|1.8|0.2|0.3|2.3|27.5|                                                               
|decode_asr_lm_lm_train_lm_transformer2_en_bpe5000_valid.loss.ave_asr_model_valid.acc.ave_use_k2_is_ctc_decoding_true_use_nbest_rescoring_true/test_other|2939|52343|95.2|4.2|0.6|0.5|5.3|44.7|

sw005320 reviewed (approved):
My comments are minor, but there seem to have several typos and missing explanations.
For the typos, you may use some spell check tools to fix them (maybe you can make some follow-up PRs to fix them?)
For the missing explanations, I'll raise them when I found them.

Thanks a lot for your great contributions!

glynpu commented:
Thanks for approving!

> maybe you can make some follow-up PRs to fix them?

Yes, I will continue to support integration of k2 into espnet.
